log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/StringBuildersTest.java

```diff
@@ -0,0 +1,90 @@
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class StringBuildersTest {
+
+    static Stream<Arguments> testTruncateLines_happyCases() {
+        return Stream.of(
+                // maxOccurrenceCount < lines count
+                Arguments.of("abc\ndef\nghi\njkl\n", "\n", 2, "abc\ndef\n"),
+                // maxOccurrenceCount == lines count
+                Arguments.of("abc|def|ghi|jkl|", "|", 4, "abc|def|ghi|jkl|"),
+                // maxOccurrenceCount > lines count
+                Arguments.of("abc|def|ghi|jkl|", "|", 10, "abc|def|ghi|jkl|"),
+                // maxOccurrenceCount ==  Integer.MAX_VALUE
+                Arguments.of("abc|def|ghi|jkl|", "|", Integer.MAX_VALUE, "abc|def|ghi|jkl|"),
+                // maxOccurrenceCount ==  0
+                Arguments.of("abc|def|ghi|jkl|", "|", 0, ""),
+                // empty buffer
+                Arguments.of("", "|", 2, ""),
+                // empty delimiter
+                Arguments.of("abc|def|ghi|jkl|", "", 2, "abc|def|ghi|jkl|"),
+                // delimiter |
+                Arguments.of("|", "|", 10, "|"),
+                Arguments.of("||", "|", 10, "||"),
+                Arguments.of("a|", "|", 10, "a|"),
+                Arguments.of("|a", "|", 10, "|"),
+                // delimiter ||
+                Arguments.of("||", "||", 10, "||"),
+                Arguments.of("|||", "||", 10, "||"),
+                Arguments.of("||||", "||", 10, "||||"),
+                Arguments.of("a|", "||", 10, ""),
+                Arguments.of("a||", "||", 10, "a||"),
+                Arguments.of("a|||", "||", 10, "a||"),
+                Arguments.of("a||||", "||", 10, "a||||"),
+                Arguments.of("|a", "||", 10, ""),
+                Arguments.of("||a", "||", 10, "||"),
+                Arguments.of("|||a", "||", 10, "||"),
+                Arguments.of("||||a", "||", 10, "||||"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("testTruncateLines_happyCases")
+    void testTruncateLinesHappyCases(String input, String delimiter, int maxOccurrenceCount, String expected) {
+        final StringBuilder buffer = new StringBuilder(input);
+        StringBuilders.truncateAfterDelimiter(buffer, delimiter, maxOccurrenceCount);
+        assertThat(buffer.toString()).isEqualTo(expected);
+    }
+
+    static Stream<Arguments> testTruncateLines_failCases() {
+        return Stream.of(
+                // negative maxOccurrenceCount
+                Arguments.of("abc\ndef\nghi\njkl\n", "\n", -1, IllegalArgumentException.class),
+                // null buffer
+                Arguments.of(null, "|", 10, NullPointerException.class),
+                // null delimiter
+                Arguments.of("abc|def|ghi|jkl|", null, 10, NullPointerException.class));
+    }
+
+    @ParameterizedTest
+    @MethodSource("testTruncateLines_failCases")
+    void testTruncateLinesFailCases(String input, String delimiter, int maxOccurrenceCount, Class<Throwable> expected) {
```

Comment:
Could you please add to these and all other variables `final` modifier, if applicable, please? (Sorry for bugging about this convention, which neither I am a fan of.)

Reply:
updated